sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/SaveAsHiveFile.scala

```diff
@@ -97,7 +99,43 @@ private[hive] trait SaveAsHiveFile extends DataWritingCommand {
       options = Map.empty)
   }
 
-  protected def getExternalTmpPath(
+  /*
+   * Mostly copied from Context.java#getMRTmpPath of Hive 2.3
+   *
+   */
+  def getMRTmpPath(
+      hadoopConf: Configuration,
+      sessionScratchDir: String,
+      scratchDir: String): Path = {
+
+    // Hive's getMRTmpPath uses nonLocalScratchPath + '-mr-10000',
+    // which is ruled by 'hive.exec.scratchdir' including file system.
+    // This is the same as Spark's #oldVersionExternalTempPath
+    // Only difference between #oldVersionExternalTempPath and Hive 2.3.0's is HIVE-7090
+    // HIVE-7090 added user_name/session_id on top of 'hive.exec.scratchdir'
+    // Here it uses session_path unless it's emtpy, otherwise uses scratchDir
+    val sessionPath = Option(sessionScratchDir).filterNot(_.isEmpty).getOrElse(scratchDir)
+    logDebug(s"session path '${sessionPath.toString}' is used")
+
+    val mrScratchDir = oldVersionExternalTempPath(new Path(sessionPath), hadoopConf, sessionPath)
+    logDebug(s"MR scratch dir '$mrScratchDir/-mr-10000' is used")
+    new Path(mrScratchDir, "-mr-10000")
+  }
+
+  private def isBlobStoragePath(path: Path): Boolean = {
+    path != null && isBlobStorageScheme(Option(path.toUri.getScheme).getOrElse(""))
```

Comment:
I wonder how Hive handles default scheme. Does it work when `fs.default.name` is set to `s3` too?

Reply:
As far as I tried, it worked even when `fs.default.name` is set to `s3.

Reply:
I believe that there are only two cases where S3 is used as scratch dir even for S3 path with `spark.sql.hive.blobstore.use.blobstore.as.scratchdir=false`.
- When `_hive.hdfs.session.path` is set to `s3`
- When `_hive.hdfs.session.path` is not set and `hive.exec.scratchdir` is set to `s3`

In both cases, these params can be configured by users, so it won't be an issue.